rfc/0006-kubewarden-versioning.md

```diff
@@ -0,0 +1,203 @@
+SDK's. This proposal is going to ignore this dimension in purpose,
+because Kubewarden runtime components (`policy-server` and `kwctl`)
+should be compatible with policies previously built, even with older
+language SDK's.
+
+### Public interfaces
+
+Our public interfaces with users are as follows:
+
+- Helm Charts
+  - `values.yaml`
+- Kubewarden Controller
+  - CLI args
+  - Custom Resource Definitions
+    - Versioned
+- Policy Server
+  - CLI args
+  - Configuration
+    - sigstore related files
+    - Usually handled by the Kubewarden Controller. Regardless we
+      should treat this as public interface, given some users might
+      run the Policy Server independent of the Kubewarden Controller,
+      and might configure it with the configuration file.
+- kwctl
+  - CLI args
+  - Configuration
+    - `sources.yaml`
+    - `docker.json`
+    - sigstore related files
+
+There is also the `waPC` contract between the `policy-evaluator`
+(consumed by the Policy Server and kwctl) and Policies and SDK's.
+
+## Kubewarden toplevel version
+
+This proposal envisions using [`CalVer`](https://calver.org/) to
+version the Kubewarden project.
+
+Had Kubewarden be released today, it would be `Kubewarden
+22.03`. This makes it trivial to understand how old your version of
+Kubewarden is, and how much your installation is drifting from the
+latest releases.
+
+## Kubewarden component versions
+
+Every component in Kubewarden might have its own versioning. The
+Kubewarden maintainers are going to maintain a list of all releases.
+
+This list of releases looks as follows:
+
+```yaml
+"22.03":
+  - component: kubewarden-crds-chart
+    version: 0.1.1
+  - component: kubewarden-controller-chart
+    version: 0.3.6
+  - component: kubewarden-controller
+    version: v0.4.5
+  - component: policy-server
+    version: v0.2.6
+  - component: kwctl
+    version: v0.2.5
+"22.05":
+  - component: kubewarden-crds-chart
+    version: 0.2.0
+  - component: kubewarden-controller-chart
+    version: 0.4.0
+  - component: kubewarden-controller
+    version: v0.4.8
+  - component: policy-server
+    version: v0.3.0
+  - component: kwctl
+    version: v0.3.1
+...
+```
+
+As more components are relevant they can be added to the list as time
+grows.
+
+This map allows us to generate documentation and a single source of
+truth to refer to when a user has a problem, or when we are debugging
+an issue. It helps to understand if any component has drifted from the
+version we are expecting.
+
+The helm chart will be responsible of storing the "Kubewarden version"
+in a resource, like a ConfigMap. So it's easy to inspect it and
+understand what "Kubewarden version" a cluster is running.
+
+Had other external components such as `kwctl` need to synchronize with
+the cluster, they can inspect this ConfigMap and realize if they are
+compatible with the running version of Kubewarden on the cluster they
+are connecting to.
```

Comment:
who creates that Configmap to track the Kubewarden version? (what happens if the user deploys the controller on its own, or policy-server, even if not officially supported? are we ok making that usecase more difficult?)

Reply:
I think the umbrella helm chart (versioned with calver) can install a configmap with the full map of versions of all components, so if needed, some components can always assume that this configmap exists in order to check their own compatibility with the rest of the components if needed.

This configmap would be autogenerated.

Reply:
> I think the umbrella helm chart (versioned with calver) can install a configmap

I don't think having an umbrella helm chart is feasible (as subcharts all get flattened and installed at the same time, giving us the problem of the controller not being up when instantiating PolicyServer). Unless it is a hypper chart.

Then, the problem is how to install it. I suppose the controller would do it.